Update heap implementation

CptGibbon opened the pull request:
This PR makes some improvements to pwndbg's heap-related commands.
### Primary changes

* The "heap" and "top_chunk" commands now print just the relevant information about chunks they encounter, rather than dumping all possible fields whether they are in use or not.

**Before:** 
<img width="370" alt="image" src="https://user-images.githubusercontent.com/16000770/77214591-2af8a680-6acd-11ea-8010-a49ff153a923.png">
<img width="370" alt="image" src="https://user-images.githubusercontent.com/16000770/77214684-7a3ed700-6acd-11ea-8f56-9a7f0dd874d9.png">

**After:**
<img width="370" alt="image" src="https://user-images.githubusercontent.com/16000770/77214785-e7526c80-6acd-11ea-8e70-aa26702452de.png">
<img width="370" alt="image" src="https://user-images.githubusercontent.com/16000770/77214793-ef121100-6acd-11ea-9c3a-98a77fed3088.png">

* Heap commands now default to using the currently debugged thread's arena, rather than the main arena. This makes things a lot more convenient when debugging multi-threaded applications and fixes the issue where no bin labels were displayed when printing a foreign thread's heap.

* Tcache commands have been placed behind an OnlyWithTcache decorator, which stops them from being invoked when debugging versions of GLIBC compiled without tcache support.

### Housekeeping:

* Some variables have been renamed for clarity; for example, a variable named "main_arena" was being used to refer to arenas that may or may not be the main arena. "main_heap" has been renamed to "allocator" because it refers to the allocator implementation rather than any specific heap.

* Argparse descriptions are single-line now so they don't get truncated by the "pwndbg" command.

* Some PEP8 amendments such as whitespace removal and docstring width.

CptGibbon committed:
Update heap implementation

* Heap commands default to using the debugged thread's arena, rather
than the main arena.

* Tcache commands are behind an OnlyWithTcache decorator.

* The "heap" & "top_chunk" commands only print relevant information
about chunks they encounter.

* Integrate _fetch_tcache_addr() into thread_cache().

* Move some code from get_region() into get_heap_boundaries().

* Rename some variables for clarity.

* Argparse descriptions are single-line so they don't get truncated.

* Some PEP8 changes such as whitespace & docstring width.

GrosQuildu commented:
I don't like new "only relevant" printing. Often its useful to quickly view all data that may matter. But the new formatting is cool. Maybe we can print all the data but mark not relevant fields in red?

disconnect3d commented:
Or have a flag like `-a` / `--all` to be able to show all data? Or/and a config variable, heh

GrosQuildu commented:
Hem, I would rather have `--relevant` ;p

CptGibbon commented:
> Hem, I would rather have `--relevant` ;p

Sure, I can make this change.
The reason I chose to print only relevant fields by default was #347, in which @zachriggle comments _"The current `ptmalloc` printers . . . print lots of fields that are not relevant"_.

CptGibbon committed:
Catch initialized heap but no thread arena case

When the main thread's arena has been initialized, but the debugged
thread's arena hasn't, default to the main arena when invoking
heap-related commands.

CptGibbon committed:
Revert how cells are unpacked in vis_heap_chunks()

This corrects an error I introduced in 798bcb6, which resulted in
vis_heap_chunks() printing negative values on the heap with a minus sign
prepended.

disconnect3d commented:
@GrosQuildu revisit pls

CptGibbon committed:
Add a --verbose flag to the heap() command

The -v or --verbose flag causes the heap() command to print every field
of every chunk it comes across, regardless of whether those fields are
in use.

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -574,11 +603,13 @@ def vis_heap_chunks(address=None, count=None, naive=None):
 def bin_ascii(bs):
     from string import printable
-    valid_chars = list(map(ord, set(printable) - set('\t\r\n')))
+    valid_chars = list(map(ord, set(printable) - set('\t\r\n\x0c')))
```
Why?

CptGibbon replied:
Without this change, in xterm I see the following effect when vis_heap_chunks() attempts to hexdump the form feed (`0x0c`) character:
![Capture](https://user-images.githubusercontent.com/16000770/80233764-dc7a8280-860b-11ea-8d94-af5d37458352.PNG)

GrosQuildu commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -134,40 +136,23 @@ def arenas(self):
+                tcache = tcache_addr
 
-        if tcache_addr is not None:
+        if tcache is not None:
```
Isn't it always not None?

CptGibbon replied:
You're correct, I'll remove this check.

GrosQuildu commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -268,20 +241,25 @@ def minsize(self):
+    @pwndbg.memoize.reset_on_objfile
+    def multithreaded(self):
+        """Is malloc operating within a multithreaded environment."""
+        addr = pwndbg.symbol.address('__libc_multiple_threads')
```
i686 libcs seems to not have this symbol. Any other way to detect multithreading?

GrosQuildu replied:
Maybe primitive like:
```py
        if addr:
            return pwndbg.memory.s32(addr) > 0
        return len(gdb.execute('info threads', to_string=True).split('\n')) > 3
```

CptGibbon replied:
I'm unfamiliar with the i686 GLIBC versions so I'll defer to you and make your suggested change.

GrosQuildu commented:
Seems fine, cool! Just solve conflict and we may merge.

Only have a question about `allocator` varname. Where it came from? `malloc.c` code? Cause name like `heap_current` would be far more readable than `allocator`.

CptGibbon committed:
Remove unnecessary if statement from thread_cache()

Remove a check for whether the 'tcache' variable was 'None', when it
could never have been 'None'.

CptGibbon committed:
Merge branch 'dev' into dev

CptGibbon commented:
> Only have a question about `allocator` varname. Where it came from? `malloc.c` code? Cause name like `heap_current` would be far more readable than `allocator`.

I picked it because the presence of files like `dlmalloc.py` in the `heap` directory implies that multiple allocators could be supported in future. I feel like `heap_current` could refer to an actual heap, which there may be lots of in a single process, whereas `allocator` is an umbrella term for all the metadata relevant to every heap & arena.

GrosQuildu commented:
Hm, ok. Only thing with the `__libc_multiple_threads` left.

CptGibbon committed:
Add alternate test for multiple threads

Some versions of GLIBC apparently don't have the __libc_multiple_threads
symbol. In these cases use GDB to determine whether there is more than
one active thread.

GrosQuildu commented:
@disconnect3d merge pls.

disconnect3d commented:
@GrosQuildu i'm giving u a push access to the repo
